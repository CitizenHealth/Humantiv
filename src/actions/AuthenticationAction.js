import firebase from "react-native-firebase";
import { GoogleSignin } from 'react-native-google-signin';
import { AccessToken, LoginManager } from 'react-native-fbsdk';

import { Actions } from "react-native-router-flux";
import { EMAIL_CHANGED,
  PASSWORD_CHANGED,
  LOGIN_USER_SUCCESS,
  LOGIN_USER_FAIL,
  LOGIN_USER,
  CREATE_USER,
  LOGOUT_USER,
  GOOGLE_LOGIN_USER,
  FACEBOOK_LOGIN_USER
} from "./types";

export const emailChanged = (text) => {
  return {
    type: EMAIL_CHANGED,
    payload: text
  };
};

export const passwordChanged = (text) => {
  return {
    type: PASSWORD_CHANGED,
    payload: text
  };
};

export const loginUser = ({ email, password }) => {
  return (dispatch) => {
    firebase.auth().signInWithEmailAndPassword(email, password)
    .then(() => {
      dispatch({ type: LOGIN_USER });
    })
    .catch(() => {
      firebase.auth().createUserWithEmailAndPassword(email, password)
      .then(() => {
        dispatch({ type: CREATE_USER });
        // Create user database
        
      })
      .catch((error) => { loginUserFail(dispatch, error); });
    });
  };
};

const loginUserSuccess = (dispatch, user) => {
  //If user not verified with email then send verification email and wait
  dispatch(
    {
      type: LOGIN_USER_SUCCESS,
      payload: user._user
    });    
  Actions.main();
  
};

const loginUserFail = (dispatch, error) => {
  dispatch(
    {
      type: LOGIN_USER_FAIL,
      payload: error.message
    });
};

export const loginGoogleUser = () => {
  return (dispatch) => {
    googleLogin(dispatch);
  };
};

export const loginFacebookUser = () => {
  return (dispatch) => {
    facebookLogin(dispatch);
  };
};

export const logoutUser = () => {
  return (dispatch) => {
    firebase.auth().signOut()
    .then(() => {
      dispatch({type: LOGOUT_USER});
    }).catch( (error) => {
      console.log("Logout: Error - "+error.message);
    });
  };
};

export const fetchUser = (user) => {
  if (!user)
    console.log("User is null");
  return (dispatch) => {
    dispatch({
      type: LOGIN_USER_SUCCESS,
      payload: user._user
    });
  };
};

// Calling this function will open Google for login.
 const googleLogin = (dispatch) => {
  // Add configuration settings here:
  return GoogleSignin.configure({
    iosClientId: "847047929311-ibked2hj6k5263b55bcjfnpiemrbpans.apps.googleusercontent.com", // only for iOS
  })
  .then(() => {
    GoogleSignin.hasPlayServices({ autoResolve: true }).then(() => {
      // play services are available. can now configure library
      GoogleSignin.signIn()
      .then((data) => {
        
        // create a new firebase credential with the token
        const credential = firebase.auth.GoogleAuthProvider.credential(data.idToken, data.accessToken);
        dispatch({ type: GOOGLE_LOGIN_USER });
  
        // login with credential
        return firebase.auth().signInWithCredential(credential);
      })
      .then((user) => {
        console.info(JSON.stringify(user.toJSON()));
        loginUserSuccess(dispatch, user);
        console.log("Signed In with Google");
      })
      .catch((error) => {
        console.error(`Login fail with error: ${error}`);
        loginUserFail(dispatch, error);
      });
    })
    .catch((err) => {
      console.log("Play services error", err.code, err.message);
    });
   });
};

// Calling the following function will open the FB login dialogue:
const facebookLogin = (dispatch) => {
  return LoginManager
    .logInWithReadPermissions(['public_profile', 'email'])
    .then((result) => {
      if (!result.isCancelled) {
        console.log(`Login success with permissions: ${result.grantedPermissions.toString()}`);
        // get the access token
        return AccessToken.getCurrentAccessToken();
      }
    })
    .then(data => {
      if (data) {
        // create a new firebase credential with the token
        const credential = firebase.auth.FacebookAuthProvider.credential(data.accessToken);
        dispatch({ type: FACEBOOK_LOGIN_USER });
        // login with credential
        return firebase.auth().signInWithCredential(credential);
      }
    })
    .then((user) => {
      if (user) {
        console.info(JSON.stringify(user.toJSON()));
        loginUserSuccess(dispatch, user);
      }
    })
    .catch((error) => {
      console.log(`Login fail with error: ${error}`);
      loginUserFail(dispatch, error);
    });
};
