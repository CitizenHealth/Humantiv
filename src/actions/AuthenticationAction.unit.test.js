import * as actions from './AuthenticationAction';

describe('nameChanged returns action', () => {
	it('returns an action', () => {
		expect(actions.nameChanged('Leo Messi')).toMatchSnapshot();
	});

	it('returns an actions if passed a null', () => {
		expect(actions.nameChanged(null)).toMatchSnapshot();
	});

	it('returns an actions if passed an empty object', () => {
		expect(actions.nameChanged('')).toMatchSnapshot();
	});
});

describe('emailChanged returns action', () => {
	it('returns an action', () => {
		expect(actions.emailChanged('leo.messi@barca.com')).toMatchSnapshot();
	});

	it('returns an actions if passed a null', () => {
		expect(actions.emailChanged(null)).toMatchSnapshot();
	});

	it('returns an actions if passed an empty object', () => {
		expect(actions.emailChanged('')).toMatchSnapshot();
	});
});

describe('passwordChanged returns action', () => {
	it('returns an action', () => {
		expect(actions.passwordChanged('leo.messi@barca.com')).toMatchSnapshot();
	});

	it('returns an actions if passed a null', () => {
		expect(actions.passwordChanged(null)).toMatchSnapshot();
	});

	it('returns an actions if passed an empty object', () => {
		expect(actions.passwordChanged('')).toMatchSnapshot();
	});
});


describe('registerUser returns action', () => {
	it('dispatches register succeeded', async () => {
		const mockDispatch = jest.fn();
		await actions.registerUser({ name: 'Leo Messi', email: 'leo.messi@barca.com', password: 'Iamth3k1n9' })(mockDispatch);
		expect(mockDispatch.mock.calls[0][0]).toMatchSnapshot();
	});
})