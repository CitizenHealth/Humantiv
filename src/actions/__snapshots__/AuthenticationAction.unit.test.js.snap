// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emailChanged returns action returns an action 1`] = `
Object {
  "payload": "leo.messi@barca.com",
  "type": "email_changed",
}
`;

exports[`emailChanged returns action returns an actions if passed a null 1`] = `
Object {
  "payload": null,
  "type": "email_changed",
}
`;

exports[`emailChanged returns action returns an actions if passed an empty object 1`] = `
Object {
  "payload": "",
  "type": "email_changed",
}
`;

exports[`nameChanged returns action returns an action 1`] = `
Object {
  "payload": "Leo Messi",
  "type": "name_changed",
}
`;

exports[`nameChanged returns action returns an actions if passed a null 1`] = `
Object {
  "payload": null,
  "type": "name_changed",
}
`;

exports[`nameChanged returns action returns an actions if passed an empty object 1`] = `
Object {
  "payload": "",
  "type": "name_changed",
}
`;

exports[`passwordChanged returns action returns an action 1`] = `
Object {
  "payload": "leo.messi@barca.com",
  "type": "password_changed",
}
`;

exports[`passwordChanged returns action returns an actions if passed a null 1`] = `
Object {
  "payload": null,
  "type": "password_changed",
}
`;

exports[`passwordChanged returns action returns an actions if passed an empty object 1`] = `
Object {
  "payload": "",
  "type": "password_changed",
}
`;

exports[`registerUser returns action dispatches register succeeded 1`] = `
Object {
  "type": "create_user",
}
`;
